airflow/models/renderedtifields.py

```diff
@@ -0,0 +1,135 @@
+        self.ti = ti
+        if render_templates:
+            ti.render_templates()
+        self.rendered_fields = {
+            field: serialize_template_field(
+                getattr(self.task, field)
+            ) for field in self.task.template_fields
+        }
+
+    def __repr__(self):
+        return f"<{self.__class__.__name__}: {self.dag_id}.{self.task_id} {self.execution_date}"
+
+    @classmethod
+    @provide_session
+    def get_templated_fields(cls, ti: TaskInstance, session: Session = None) -> Optional[dict]:
+        """
+        Get templated field for a TaskInstance from the RenderedTaskInstanceFields
+        table.
+
+        :param ti: Task Instance
+        :param session: SqlAlchemy Session
+        :return: Rendered Templated TI field
+        """
+        result = session.query(cls.rendered_fields).filter(
+            cls.dag_id == ti.dag_id,
+            cls.task_id == ti.task_id,
+            cls.execution_date == ti.execution_date
+        ).one_or_none()
+
+        if result:
+            rendered_fields = result.rendered_fields
+            return rendered_fields
+        else:
+            return None
+
+    @provide_session
+    def write(self, session: Session = None):
+        """Write instance to database
+
+        :param session: SqlAlchemy Session
+        """
+        session.merge(self)
+
+    @classmethod
+    @provide_session
+    def delete_old_records(
+        cls, task_id: str, dag_id: str,
+        num_to_keep=conf.getint("core", "max_num_rendered_ti_fields_per_task", fallback=0),
+        session: Session = None
+    ):
+        """
+        Keep only Last X (num_to_keep) number of records for a task by deleting others
+
+        :param task_id: Task ID
+        :param dag_id: Dag ID
+        :param num_to_keep: Number of Records to keep
+        :param session: SqlAlchemy Session
+        """
+        if num_to_keep > 0:
+            # Fetch Top X records given dag_id & task_id ordered by Execution Date
+            subq1 = (
+                session
+                .query(cls.dag_id, cls.task_id, cls.execution_date)
+                .filter(cls.dag_id == dag_id, cls.task_id == task_id)
+                .order_by(cls.execution_date.desc())
+                .limit(num_to_keep)
+                .subquery('subq1')
+            )
+
+            # Second Subquery
+            # Workaround for MySQL Limitation (https://stackoverflow.com/a/19344141/5691525)
+            # Limitation: This version of MySQL does not yet support
+            # LIMIT & IN/ALL/ANY/SOME subquery
+            subq2 = (
+                session
+                .query(subq1.c.dag_id, subq1.c.task_id, subq1.c.execution_date)
+                .subquery('subq2')
+            )
+
+            session.query(cls) \
+                .filter(and_(
+                    cls.dag_id == dag_id,
+                    cls.task_id == task_id,
+                    tuple_(cls.dag_id, cls.task_id, cls.execution_date).notin_(subq2))) \
```

Comment:
Tuple is problematic. It is not supported by all RDMS. 
```
The composite IN construct is not supported by all backends, and is
            currently known to work on PostgreSQL, MySQL, and SQLite.
            Unsupported backends will raise a subclass of
            :class:`~sqlalchemy.exc.DBAPIError` when such an expression is
            invoked.
``` We don't use this syntax elsewhere. Do you really need to use it here?

Reply:
It works for mysql and Postgres, and sqlite doesn't it?

Reply:
Yes, as the docs you quoted say :facepalm:

Reply:
Hmm are we planning to support all the backends supported by Sqlalchemy?

Reply:
I will create a new issue to sort this out

Reply:
Created at https://issues.apache.org/jira/browse/AIRFLOW-7045